lib/puppet/resource/capability_finder.rb

```diff
@@ -22,6 +26,29 @@ def self.find(environment, code_id, cap)
       raise Puppet::DevError, 'PuppetDB is not available'
     end
 
+    resources = search(environment, nil, cap)
+
+    if resources.size > 1 && code_id
+      Puppet.debug "Found multiple resources when looking up capability #{cap}, filtering by code id #{code_id}"
+      resources = search(environment, code_id, cap)
+    end
+
+    if resources.size > 1
+      raise Puppet::DevError,
+        "Unexpected response from PuppetDB when looking up #{cap}:\n" +
+        "expected exactly one resource but got #{resources.size};\n" +
+        "returned data is:\n#{resources.inspect}"
```

Comment:
Picky, it is better to use a `\` instead of a `+` as the `\` is handled by the ruby parser (creating a single string) vs. the `+` which creates individual strings that are then concatenated,